File: src/test/java/co/nums/intellij/aem/htl/file/HtlLanguageSubstitutorTest.java

package co.nums.intellij.aem.htl.file;

import co.nums.intellij.aem.constants.JcrConstants;
import co.nums.intellij.aem.htl.HtlLanguage;
import com.intellij.ide.highlighter.HtmlFileType;
import com.intellij.lang.Language;
import com.intellij.openapi.fileTypes.FileType;
import com.intellij.openapi.fileTypes.MockLanguageFileType;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;
import org.junit.jupiter.params.provider.ValueSource;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

class HtlLanguageSubstitutorTest {

	private static final Project ANY_PROJECT = mock(Project.class);
	private static final FileType NON_HTL_FILE_TYPE = MockLanguageFileType.INSTANCE;
	private static final VirtualFile JCR_ROOT_DIR = directoryWithName(JcrConstants.JCR_ROOT_DIRECTORY_NAME);

	private HtlLanguageSubstitutor sut;

	private VirtualFile file;

	@BeforeEach
	void setUp() {
		sut = new HtlLanguageSubstitutor();
		file = mock(VirtualFile.class);
	}

	@ParameterizedTest
	@ValueSource(strings = {
			//@formatter:off
			"jcr_root",
			"jcr_root/apps",
			"jcr_root/apps/components",
			"home/user/aem-project/test-module/jcr_root",
			"home/user/aem-project/test-module/jcr_root/apps",
			"home/user/aem-project/test-module/jcr_root/apps/components"
			//@formatter:on
	})
	void shouldSubstituteHtmlFileInJcrRootDirectory(String path) {
		when(file.getFileType()).thenReturn(HtmlFileType.INSTANCE);
		prepareFileDirectoryPath(file, path);

		Language language = sut.getLanguage(file, ANY_PROJECT);

		assertThat(language).isEqualTo(HtlLanguage.INSTANCE);
	}

	@Test
	void shouldNotSubstituteNonHtmlFileInJcrRootDirectory() {
		when(file.getFileType()).thenReturn(NON_HTL_FILE_TYPE);
		when(file.getParent()).thenReturn(JCR_ROOT_DIR);

		Language language = sut.getLanguage(file, ANY_PROJECT);

		assertThat(language).isNull();
	}

	private static String[] pathsOutOfJcrRoot() {
		return new String[] {
				//@formatter:off
				"not_jcr_root",
				"jcr_root_not",
				"_jcr_root",
				"jcr_root_",
				"jcrroot",
				"not_jcr_root/apps",
				"jcr_root_not/apps",
				"_jcr_root/apps",
				"jcr_root_/apps",
				"jcrroot/apps",
				"home/user/aem-project/test-module/not_jcr_root",
				"home/user/aem-project/test-module/jcr_root_not",
				"home/user/aem-project/test-module/_jcr_root",
				"home/user/aem-project/test-module/jcr_root_",
				"home/user/aem-project/test-module/jcrroot"
				//@formatter:on
		};
	}

	@ParameterizedTest
	@MethodSource("pathsOutOfJcrRoot")
	void shouldNotSubstituteHtmlFileOutOfJcrRootDirectory(String pathOutOfJcrRoot) {
		when(file.getFileType()).thenReturn(HtmlFileType.INSTANCE);
		prepareFileDirectoryPath(file, pathOutOfJcrRoot);

		Language language = sut.getLanguage(file, ANY_PROJECT);

		assertThat(language).isNull();
	}

	@ParameterizedTest
	@MethodSource("pathsOutOfJcrRoot")
	void shouldNotSubstituteNonHtmlFileOutOfJcrRootDirectory(String pathOutOfJcrRoot) {
		when(file.getFileType()).thenReturn(NON_HTL_FILE_TYPE);
		prepareFileDirectoryPath(file, pathOutOfJcrRoot);

		Language language = sut.getLanguage(file, ANY_PROJECT);

		assertThat(language).isNull();
	}

	private static void prepareFileDirectoryPath(VirtualFile file, String path) {
		String[] pathParts = path.split("/");
		VirtualFile currentFile = file;
		for (int i = pathParts.length - 1; i >= 0; i--) {
			VirtualFile parentDirectory = directoryWithName(pathParts[i]);
			when(currentFile.getParent()).thenReturn(parentDirectory);
			currentFile = parentDirectory;
		}
	}

	private static VirtualFile directoryWithName(String name) {
		VirtualFile directory = mock(VirtualFile.class);
		when(directory.getName()).thenReturn(name);
		return directory;
	}

}

File: src/test/kotlin/co/nums/intellij/aem/htl/file/HtlLanguageSubstitutorTest.kt
package co.nums.intellij.aem.htl.file

import co.nums.intellij.aem.htl.HtlLanguage
import co.nums.intellij.aem.service.JcrRoots
import co.nums.intellij.aem.settings.AemSettings
import co.nums.intellij.aem.test.util.whenGetService
import com.intellij.lang.Language
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VirtualFile
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.converter.*
import org.junit.jupiter.params.provider.CsvSource
import org.mockito.Mockito
import org.mockito.Mockito.`when`
import org.mockito.Mockito.mock


class HtlLanguageSubstitutorTest {

    private lateinit var project: Project
    private lateinit var aemSettings: AemSettings
    private lateinit var jcrRoots: JcrRoots
    private lateinit var testFile: VirtualFile

    private lateinit var sut: HtlLanguageSubstitutor

    @BeforeEach
    fun setUp() {
        project = mock(Project::class.java, Mockito.RETURNS_DEEP_STUBS)
        aemSettings = mock(AemSettings::class.java)
        jcrRoots = mock(JcrRoots::class.java)
        whenGetService<AemSettings>(project).thenReturn(aemSettings)
        whenGetService<JcrRoots>(project).thenReturn(jcrRoots)
        testFile = mock(VirtualFile::class.java)
        sut = HtlLanguageSubstitutor()
    }

    @ParameterizedTest(name = "should return {3} for: aemSupportEnabled={0}, htmlFile={1}, fileUnderJcrRoot={2}")
    @CsvSource(
            "true ,true ,true ,HtlLanguage",
            "true ,true ,false,null",
            "true ,false,true ,null",
            "true ,false,false,null",
            "false ,true ,true ,null",
            "false ,true ,false,null",
            "false ,false,true ,null",
            "false ,false,false,null"
    )
    fun testSubstitution(aemSupportEnabled: Boolean, htmlFile: Boolean, fileUnderJcrRoot: Boolean, @ConvertWith(LanguageArgumentConverter::class) expectedLanguage: Language?) {
        `when`(aemSettings.aemSupportEnabled).thenReturn(aemSupportEnabled)
        `when`(testFile.extension).thenReturn(if (htmlFile) "html" else "txt")
        `when`(jcrRoots.contains(testFile)).thenReturn(fileUnderJcrRoot)

        val actualLanguage = sut.getLanguage(testFile, project)

        assertThat(actualLanguage).isEqualTo(expectedLanguage)
    }

    class LanguageArgumentConverter : SimpleArgumentConverter() {
        override fun convert(source: Any?, targetType: Class<*>) = if (source == "HtlLanguage") HtlLanguage else null
    }

}
